posthog/hogql_queries/insights/retention_query_runner.py

```diff
@@ -0,0 +1,316 @@
+        start_of_interval_sql = self.query_date_range.get_start_of_interval_hogql(
+            source=ast.Field(chain=["events", "timestamp"])
+        )
+
+        if event_query_type == RetentionQueryType.TARGET_FIRST_TIME:
+            event_date_expr = ast.Call(name="min", args=[start_of_interval_sql])
+        else:
+            event_date_expr = start_of_interval_sql
+
+        fields = [
+            ast.Alias(alias="event_date", expr=event_date_expr),
+            ast.Alias(alias="target", expr=ast.Field(chain=["events", "person_id"])),
+        ]
+
+        if event_query_type in [RetentionQueryType.TARGET, RetentionQueryType.TARGET_FIRST_TIME]:
+            source_timestamp = ast.Field(chain=["events", "timestamp"])
+            if event_query_type == RetentionQueryType.TARGET_FIRST_TIME:
+                source_timestamp = ast.Call(
+                    name="min",
+                    args=[source_timestamp],
+                )
+
+            datediff_call = ast.Call(
+                name="dateDiff",
+                args=[
+                    ast.Constant(value=self.query_date_range.interval_name),
+                    self.query_date_range.get_start_of_interval_hogql(),
+                    self.query_date_range.get_start_of_interval_hogql(
+                        source=source_timestamp,
+                    ),
+                ],
+            )
+            fields.append(
+                ast.Alias(alias="breakdown_values", expr=ast.Array(exprs=[datediff_call])),
+            )
+
+        event_filters = [
+            entity_to_expr(
+                entity=self.target_entity
+                if event_query_type in [RetentionQueryType.TARGET, RetentionQueryType.TARGET_FIRST_TIME]
+                else self.returning_entity
+            )
+        ]
+
+        if self.query.properties is not None and self.query.properties != []:
+            event_filters.append(property_to_expr(self.query.properties, self.team))
+
+        if (
+            self.query.filterTestAccounts
+            and isinstance(self.team.test_account_filters, list)
+            and len(self.team.test_account_filters) > 0
+        ):
+            for prop in self.team.test_account_filters:
+                event_filters.append(property_to_expr(prop, self.team))
+
+        date_filter_expr = self.date_filter_expr(event_query_type)
+        if event_query_type != RetentionQueryType.TARGET_FIRST_TIME:
+            event_filters.append(date_filter_expr)
+
+        group_by_fields = None
+        having_expr = None
+        if event_query_type == RetentionQueryType.TARGET_FIRST_TIME:
+            group_by_fields = [ast.Field(chain=["target"])]
+            having_expr = date_filter_expr
+        elif event_query_type == RetentionQueryType.RETURNING:
+            group_by_fields = [ast.Field(chain=["target"]), ast.Field(chain=["event_date"])]
+
+        result = parse_select("SELECT * FROM events", timings=self.timings)
```

Comment:
Nit: you can probably get by here by just writing the AST manually. Feel free to ignore, as this is fine too.

Reply:
Yeah I had attempted something like that and it didn't work here. Will find out what it was.